text/0100-xcm-multi-type-asset-transfer.md

```diff
@@ -0,0 +1,279 @@
+for fees on `dest` chain. This single asset shall be remotely handled/received by the **first instruction** in the
+onward XCM and shall be followed by a `BuyExecution` instruction using it.
+If `remote_fees` is set to `None`, the **first instruction** in the onward XCM shall be a `UnpaidExecution` instruction.
+The rest of the assets shall be handled by subsequent instructions, thus also finally allowing
+[single asset buy execution](https://github.com/paritytech/polkadot-sdk/issues/2423) barrier security recommendation.
+
+The `BuyExecution` appended to the onward XCM specifies `WeightLimit::Unlimited`, thus being limited only by the
+`remote_fees` asset "amount". This is a deliberate decision for enhancing UX - in practice, people/dApps care about
+limiting the amount of fee asset used and not the actually used weight.
+
+The onward XCM, following the assets transfers instructions, `ClearOrigin` or `DescendOrigin` instructions shall be
+appended to stop acting on behalf of the source chain, then the caller-provided `remote_xcm` shall also be appended,
+allowing the caller to control what to do with the transferred assets.
+
+### Example usage: transferring 2 different asset types across 3 chains
+
+- Transferring ROCs as the native asset of `RococoAssetHub` and PENs as the native asset of `Penpal`,
+- Transfer origin is `Penpal` (on Rococo) and the destination is `WestendAssetHub` (across the bridge),
+- ROCs are native to `RococoAssetHub` and are registered as trust-backed assets on `Penpal` and `WestendAssetHub`,
+- PENs are native to `Penpal` and are registered as teleportable assets on `RococoAssetHub` and as
+  foreign assets on `WestendAssetHub`,
+- Fees on `RococoAssetHub` and `WestendAssetHub` are paid using ROCs.
+
+We can transfer them from `Penpal` (Rococo), through `RococoAssetHub`, over the bridge to `WestendAssetHub`
+by executing a _single_ XCM message, even though we'll be mixing multiple types of transfers along the path:
+1. 1st leg of the transfer: Penpal -> Rococo Asset Hub:
+   - teleport PENs
+   - reserve withdraw ROCs
+2. 2nd leg of the transfer: Rococo Asset Hub -> Westend Asset Hub:
+   - reserve deposit both PENs and ROCs
+
+```rust
+Penpal::execute_with(|| {
+    let rocs: Asset = (rocs_id.clone(), rocs_amount).into();
+    let pens: Asset = (pens_id, pens_amount).into();
+    let assets: Assets = vec![rocs.clone(), pens.clone()].into();
+
+    // XCM to be executed at dest (Westend Asset Hub)
+    let xcm_on_dest =
+        Xcm(vec![DepositAsset { assets: Wild(All), beneficiary: beneficiary.clone() }]);
+
+    // XCM to be executed at Rococo Asset Hub
+    let context = PenpalUniversalLocation::get();
+    let reanchored_assets = assets.clone().reanchored(&local_asset_hub, &context).unwrap();
+    let reanchored_dest = destination.clone().reanchored(&local_asset_hub, &context).unwrap();
+    let reanchored_rocs_id = rocs_id.clone().reanchored(&local_asset_hub, &context).unwrap();
```

Comment:
`destination` and `rocs_id` are not defined in this example

Reply:
I added definitions for `destination` and `rocs_id` for clarity, but please note this is meant as a conceptual example. To actually compile it, it needs much more boiler-plate code..